core/src/client/importer.rs

```diff
@@ -365,16 +369,21 @@ impl Importer {
     /// This is triggered by a message coming from a header queue when the header is ready for insertion
     pub fn import_verified_headers(&self, client: &Client) -> usize {
         let max_headers_to_import = 256;
+        let lock = self.import_lock.lock();
+        let headers = self.header_queue.drain(max_headers_to_import);
+        self.import_headers(headers, client, &lock)
+    }
 
-        let _lock = self.import_lock.lock();
+    pub fn import_headers(&self, headers: Vec<Header>, client: &Client, _importer_lock: &MutexGuard<()>) -> usize {
```

Comment:
Do we need to pass a lock to this function?

Reply:
To make the caller not forget to hold the `import_lock`?

Reply:
How about using `ReentrantMutex` then?

Reply:
I want to avoid to use the recursive lock if possible.
http://www.zaval.org/resources/library/butenhof1.html

Reply:
Sounds reasonable.